invoice/src/main/java/org/killbill/billing/invoice/generator/InvoiceWithMetadata.java

```diff
@@ -80,6 +83,73 @@ public boolean apply(final InvoiceItem input) {
         });
     }
 
+    public Set<TrackingRecordId> getTrackingIds() {
+        return trackingIds;
+    }
+
+    public static class TrackingRecordId {
+        private final String trackingId;
+        private final UUID invoiceId;
+        private final UUID subscriptionId;
+        private final String unitType;
+        private final LocalDate recordDate;
+
+        public TrackingRecordId(final String trackingId, final UUID invoiceId, final UUID subscriptionId, final String unitType, final LocalDate recordDate) {
+            this.trackingId = trackingId;
+            this.invoiceId = invoiceId;
+            this.subscriptionId = subscriptionId;
+            this.unitType = unitType;
+            this.recordDate = recordDate;
+        }
+
+        public String getTrackingId() {
+            return trackingId;
+        }
+
+        public UUID getInvoiceId() {
+            return invoiceId;
+        }
+
+        public UUID getSubscriptionId() {
+            return subscriptionId;
+        }
+
+        public LocalDate getRecordDate() {
+            return recordDate;
+        }
+
+        public String getUnitType() {
+            return unitType;
+        }
+
+        @Override
+        public boolean equals(final Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (!(o instanceof TrackingRecordId)) {
+                return false;
+            }
+            final TrackingRecordId that = (TrackingRecordId) o;
+            // !!! Exclude invoiceId on purpose.
+            //
+            // The Set methods (Sets.difference) is used to exclude usage record already invoiced (on a specified invoiceId),
```

Comment:
In the past, for such cases, we did not rely on the native methods but introduced our own (see `InvoiceItem#matches` for instance). In other words, should we keep `equals` as-is and implement our custom `Sets.difference` utility function calling a different method?

I'm just worried about future automatic refactoring. Also, it could introduce other weird behaviors when we implicitly truly expect a well-behaved implementation (for instance, doesn't `HashSet` need a real hash function for the `add` operation?).